src/violation.rs

```diff
@@ -42,16 +51,38 @@ impl<'a> ViolationBuilder<'a> {
         self.fix = Some(Some(Rc::new(callback)));
         self
     }
+
+    pub fn message(&mut self, message: impl Into<String>) -> &mut Self {
+        let message = message.into();
+        self.message_or_message_id = Some(MessageOrMessageId::Message(message));
+        self
+    }
+
+    pub fn message_id(&mut self, message_id: impl Into<String>) -> &mut Self {
+        let message_id = message_id.into();
+        self.message_or_message_id = Some(MessageOrMessageId::MessageId(message_id));
+        self
+    }
 }
 
-#[derive(Clone)]
+#[derive(Clone, Debug)]
+pub enum MessageOrMessageId {
+    Message(String),
+    MessageId(String),
+}
+
+pub type ViolationData = HashMap<String, String>;
```

Comment:
No idea if (if eg wanting to mimic ESLint) these "data" values would want to be something "broader" than `String` (eg maybe `serde_json::Value`)?